### PR TITLE
feat(dango): oracle contract

### PR DESCRIPTION
**Pull request by larry0x**
the way this oracle contract works:

- it is to be defined as a cronjob. let's call the time interval it's execute the **epoch** (say, every 30 seconds)
- during an epoch, a set of **guardians** submit prices
- at the end of each epoch, the oracle contract computes the median price of each denom

a few remaining issues:

- add exponential moving average (EMA)
- how to deal with AMM LP tokens? compute them onchain or have guardians submit the prices as well?
- chain owner should be able to add/remove guardians
- handle the case of possible chain halt - oracle should exclude feeds that are too old

### CODE REVIEW

**Comment by larry0x** [comment]
an alternative solution is to use pyth. simply submit pyth data signed by wormhole guardian set.

let me research how much effort is needed for that.

File: dango/oracle/src/execute.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+#[cfg_attr(not(feature = "library"), grug::export)]
+pub fn cron_execute(ctx: SudoCtx) -> StdResult<Response> {
+    let cfg = CONFIG.load(ctx.storage)?;
+
+    let mut denom_feeds = BTreeMap::<Denom, Vec<Udec128>>::new();
+
+    // Load all feeds submitted by guardians since the last cron execute.
+    for res in PRICE_FEEDS.prefix_range(ctx.storage, None, None, Order::Ascending) {
```
Comment (apollo-sturdy) [inline dango/oracle/src/execute.rs:51]: Why `prefix_range` instead of `range`?

File: dango/oracle/src/execute.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        if num_feeds < cfg.quorum as usize {
+            continue;
+        }
+
+        // The price feeds must be sorted in order to find the median.
+        feeds.sort();
+
+        // Find the median price.
+        // Here as an optimization, we use bitwise operators (`&` and `>>`)
+        // instead of division or modulo.
```
Comment (apollo-sturdy) [inline dango/oracle/src/execute.rs:76]: Could add these to a trait to have better clarity of code? E.g. `fn is_even`

**Comment by larry0x** [comment]
Replaced by #284

**Pull request closed without merging**